### ISSUE TITLE
Unable to run Flink Example (PortableRunner)

### ISSUE THREAD

**Issue by vaskozl** [tensorflow/tfx#2035]
I'm attempting to run the taxi_example on Apache Flink as instructed by the example using`setup_beam_on_flink.sh`, but when running the pipeline it fails almost instantly. I've also tried running the with the jobserver container from https://hub.docker.com/r/apache/beam_flink1.9_job_server to no avail.

Using tfx 0.22.0, tensorflow 2.2.0 and all recommended dependencies on pypi.org

I'm running `setup_beam_on_flink.sh` with the cherry-pick from https://github.com/tensorflow/tfx/issues/1995 as `setup_beam_on_flink.sh` will not compile the jobserver otherwise.

Running in a debian buster container. The flink jobserver log:
```
(my_env) root@5561e75bd72d:/tmp/beam/beam#  ./gradlew :runners:flink:1.9:job-server:runShadow -PflinkMasterUrl=localhost:8081
Configuration on demand is an incubating feature.

> Task :runners:flink:1.9:job-server:runShadow
Listening for transport dt_socket at address: 5005
[main] INFO org.apache.beam.runners.jobsubmission.JobServerDriver - ArtifactStagingService started on localhost:8098
[main] INFO org.apache.beam.runners.jobsubmission.JobServerDriver - Java ExpansionService started on localhost:8097
[main] INFO org.apache.beam.runners.jobsubmission.JobServerDriver - JobService started on localhost:8099
[grpc-default-executor-1] ERROR org.apache.beam.runners.jobsubmission.InMemoryJobService - Encountered Unexpected Exception for Preparation job_79743136-9488-414f-847c-d6f9bf39f838
java.lang.RuntimeException: Artifact dependencies provided but not staged for ref_Environment_default_environment_1
        at org.apache.beam.runners.jobsubmission.InMemoryJobService.resolveDependencies(InMemoryJobService.java:297)
        at org.apache.beam.runners.jobsubmission.InMemoryJobService.run(InMemoryJobService.java:244)
        at org.apache.beam.model.jobmanagement.v1.JobServiceGrpc$MethodHandlers.invoke(JobServiceGrpc.java:961)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:172)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.Contexts$ContextualizedServerCallListener.onHalfClose(Contexts.java:86)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:331)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:817)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
<============-> 98% EXECUTING [2m 13s]
> IDLE> :runners:flink:1.9:job-server:runShadow
> IDLE
> IDLE
```

The pipeline log:
```
airflow@5561e75bd72d:~/dags$ python taxi_pipeline.py | tee ~/pipeline.log
/usr/local/lib/python3.7/site-packages/tfx/orchestration/data_types.py:191: UserWarning: RuntimeParameter is only supported on KubeflowDagRunner currently.
  warnings.warn('RuntimeParameter is only supported on KubeflowDagRunner '
INFO:absl:Using 4 process(es) for Beam pipeline execution.
INFO:absl:Component ResolverNode.latest_blessed_model_resolver depends on [].
INFO:absl:Component ResolverNode.latest_blessed_model_resolver is scheduled.
INFO:absl:Component CsvExampleGen depends on [].
INFO:absl:Component CsvExampleGen is scheduled.
INFO:absl:Component StatisticsGen depends on ['Run[CsvExampleGen]'].
INFO:absl:Component StatisticsGen is scheduled.
INFO:absl:Component SchemaGen depends on ['Run[StatisticsGen]'].
INFO:absl:Component SchemaGen is scheduled.
INFO:absl:Component ExampleValidator depends on ['Run[SchemaGen]', 'Run[StatisticsGen]'].
INFO:absl:Component ExampleValidator is scheduled.
INFO:absl:Component Transform depends on ['Run[SchemaGen]', 'Run[CsvExampleGen]'].
INFO:absl:Component Transform is scheduled.
INFO:absl:Component Trainer depends on ['Run[SchemaGen]', 'Run[Transform]'].
INFO:absl:Component Trainer is scheduled.
INFO:absl:Component Evaluator depends on ['Run[Trainer]', 'Run[CsvExampleGen]', 'Run[ResolverNode.latest_blessed_model_resolver]'].
INFO:absl:Component Evaluator is scheduled.
INFO:absl:Component Pusher depends on ['Run[Evaluator]', 'Run[Trainer]'].
INFO:absl:Component Pusher is scheduled.
INFO:absl:Component ResolverNode.latest_blessed_model_resolver is running.
INFO:absl:Running driver for ResolverNode.latest_blessed_model_resolver
INFO:absl:MetadataStore with DB connection initialized
INFO:absl:Running publisher for ResolverNode.latest_blessed_model_resolver
INFO:absl:MetadataStore with DB connection initialized
INFO:absl:Component ResolverNode.latest_blessed_model_resolver is finished.
INFO:absl:Component CsvExampleGen is running.
INFO:absl:Running driver for CsvExampleGen
INFO:absl:MetadataStore with DB connection initialized
INFO:absl:select span = None
INFO:absl:Running executor for CsvExampleGen
WARNING:apache_beam.options.pipeline_options:Discarding unparseable args: ['--parallelism=4', '--execution_mode_for_batch=BATCH_FORCED']
INFO:absl:Attempting to infer TFX Python dependency for beam
INFO:absl:Copying all content from install dir /usr/local/lib/python3.7/site-packages/tfx to temp dir /tmp/tmpcbb2umtn/build/tfx
INFO:absl:Generating a temp setup file at /tmp/tmpcbb2umtn/build/tfx/setup.py
INFO:absl:Creating temporary sdist package, logs available at /tmp/tmpcbb2umtn/build/tfx/setup.log
INFO:absl:Added --extra_package=/tmp/tmpcbb2umtn/build/tfx/dist/tfx_ephemeral-0.22.0.tar.gz to beam args
INFO:absl:Generating examples.
INFO:absl:Processing input csv data /usr/local/airflow/taxi/data/simple/* to TFExample.
WARNING:root:Make sure that locally built Python SDK docker image has Python 3.7 interpreter.
Traceback (most recent call last):
  File "apache_beam/runners/common.py", line 961, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 554, in apache_beam.runners.common.SimpleInvoker.invoke_process
  File "/usr/local/lib/python3.7/site-packages/tfx/orchestration/beam/beam_dag_runner.py", line 84, in process
    self._run_component()
  File "/usr/local/lib/python3.7/site-packages/tfx/orchestration/beam/beam_dag_runner.py", line 88, in _run_component
    self._component_launcher.launch()
  File "/usr/local/lib/python3.7/site-packages/tfx/orchestration/launcher/base_component_launcher.py", line 205, in launch
    execution_decision.exec_properties)
  File "/usr/local/lib/python3.7/site-packages/tfx/orchestration/launcher/in_process_component_launcher.py", line 67, in _run_executor
    executor.Do(input_dict, output_dict, exec_properties)
  File "/usr/local/lib/python3.7/site-packages/tfx/components/example_gen/base_example_gen_executor.py", line 240, in Do
    artifact_utils.get_split_uri(output_dict['examples'], split_name)))
  File "/usr/local/lib/python3.7/site-packages/apache_beam/pipeline.py", line 524, in __exit__
    self.run().wait_until_finish()
  File "/usr/local/lib/python3.7/site-packages/apache_beam/pipeline.py", line 497, in run
    self._options).run(False)
  File "/usr/local/lib/python3.7/site-packages/apache_beam/pipeline.py", line 510, in run
    return self.runner.run_pipeline(self, self._options)
  File "/usr/local/lib/python3.7/site-packages/apache_beam/runners/portability/portable_runner.py", line 406, in run_pipeline
    job_service_handle.submit(proto_pipeline)
  File "/usr/local/lib/python3.7/site-packages/apache_beam/runners/portability/portable_runner.py", line 108, in submit
    return self.run(prepare_response.preparation_id, retrieval_token)
  File "/usr/local/lib/python3.7/site-packages/apache_beam/runners/portability/portable_runner.py", line 234, in run
    preparation_id=preparation_id, retrieval_token=retrieval_token))
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 826, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 729, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.INTERNAL
	details = ""
	debug_error_string = "{"created":"@1592928929.935875342","description":"Error received from peer ipv4:127.0.0.1:8099","file":"src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"","grpc_status":13}"
>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "taxi_pipeline.py", line 228, in <module>
    worker_parallelism=parallelism))
  File "/usr/local/lib/python3.7/site-packages/tfx/orchestration/beam/beam_dag_runner.py", line 165, in run
    absl.logging.info('Component %s is scheduled.', component_id)
  File "/usr/local/lib/python3.7/site-packages/apache_beam/pipeline.py", line 524, in __exit__
    self.run().wait_until_finish()
  File "/usr/local/lib/python3.7/site-packages/apache_beam/pipeline.py", line 497, in run
    self._options).run(False)
** Cutting Traceback **
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 729, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
RuntimeError: grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.INTERNAL
	details = ""
	debug_error_string = "{"created":"@1592928929.935875342","description":"Error received from peer ipv4:127.0.0.1:8099","file":"src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"","grpc_status":13}"
> [while running 'Run[CsvExampleGen]']
```

**Comment by ibzib** [comment]
Which Beam version are you using for the job server, and which Beam version the `apache-beam` Python package?

**Comment by Vaskozl** [comment]
```
# pip list | grep apache-beam
apache-beam                2.21.0
```
And the  the same version of beam in `setup_beam.sh` from the tfx `0.22.0` release:
https://github.com/tensorflow/tfx/blob/r0.22.0/tfx/examples/chicago_taxi_pipeline/setup/setup_beam.sh
```
Installing Beam from source code at https://github.com/apache/beam Branch release-2.21.0
```
I see now the latest script has been updated to install beam 2.22.0 so I tested with both the script and apache-beam set to `2.22.0` and still appear to get the same log with:
```
java.lang.RuntimeException: Artifact dependencies provided but not staged for ref_Environment_default_environment_1
```

**Comment by Vaskozl** [comment]
I think my original log and error (which I still received on both python and beam version 2.22.0) was using python apache-beam `2.21` and the jobsever `2.2.0` as I had picked up the latest commit  without noticing.

Making sure I'm using 2.21.0 for both, while using the cherry-pick from the related issue I still can not get the example to run:

```
root@0c2e209d12a8:/usr/local/airflow/dags# python taxi_pipeline_portable_beam.py 
/usr/local/lib/python3.7/site-packages/tfx/orchestration/data_types.py:191: UserWarning: RuntimeParameter is only supported on KubeflowDagRunner currently.  warnings.warn('RuntimeParameter is only supported on KubeflowDagRunner '
INFO:absl:Using 4 process(es) for Beam pipeline execution.
INFO:absl:Component CsvExampleGen depends on [].
INFO:absl:Component CsvExampleGen is scheduled.
INFO:absl:Component ResolverNode.latest_blessed_model_resolver depends on [].
INFO:absl:Component ResolverNode.latest_blessed_model_resolver is scheduled.
INFO:absl:Component StatisticsGen depends on ['Run[CsvExampleGen]'].
INFO:absl:Component StatisticsGen is scheduled.
INFO:absl:Component SchemaGen depends on ['Run[StatisticsGen]'].
INFO:absl:Component SchemaGen is scheduled.
INFO:absl:Component Transform depends on ['Run[CsvExampleGen]', 'Run[SchemaGen]'].
INFO:absl:Component Transform is scheduled.
INFO:absl:Component ExampleValidator depends on ['Run[StatisticsGen]', 'Run[SchemaGen]'].
INFO:absl:Component ExampleValidator is scheduled.
INFO:absl:Component Trainer depends on ['Run[Transform]', 'Run[SchemaGen]'].
INFO:absl:Component Trainer is scheduled.
INFO:absl:Component Evaluator depends on ['Run[Trainer]', 'Run[CsvExampleGen]', 'Run[ResolverNode.latest_blessed_model_resolver]'].
INFO:absl:Component Evaluator is scheduled.
INFO:absl:Component Pusher depends on ['Run[Trainer]', 'Run[Evaluator]'].
INFO:absl:Component Pusher is scheduled.
INFO:absl:Component ResolverNode.latest_blessed_model_resolver is running.
INFO:absl:Running driver for ResolverNode.latest_blessed_model_resolver
INFO:absl:MetadataStore with DB connection initializedINFO:absl:Running publisher for ResolverNode.latest_blessed_model_resolver
INFO:absl:MetadataStore with DB connection initialized
INFO:absl:Component ResolverNode.latest_blessed_model_resolver is finished.
INFO:absl:Component CsvExampleGen is running.
INFO:absl:Running driver for CsvExampleGen
INFO:absl:MetadataStore with DB connection initialized
INFO:absl:select span = None
INFO:absl:Running executor for CsvExampleGen
WARNING:apache_beam.options.pipeline_options:Discarding unparseable args: ['--parallelism=4', '--execution_mode_for_batch=BATCH_FORCED']
INFO:absl:Attempting to infer TFX Python dependency for beam
INFO:absl:Copying all content from install dir /usr/local/lib/python3.7/site-packages/tfx to temp dir /tmp/tmpgx66qcho/build/tfxINFO:absl:Generating a temp setup file at /tmp/tmpgx66qcho/build/tfx/setup.py
INFO:absl:Creating temporary sdist package, logs available at /tmp/tmpgx66qcho/build/tfx/setup.log
INFO:absl:Added --extra_package=/tmp/tmpgx66qcho/build/tfx/dist/tfx_ephemeral-0.22.0.tar.gz to beam argsINFO:absl:Generating examples.INFO:absl:Processing input csv data /root/taxi/data/simple/* to TFExample.
WARNING:root:Make sure that locally built Python SDK docker image has Python 3.7 interpreter.
ERROR:root:org.apache.flink.runtime.rest.util.RestClientException: [org.apache.flink.runtime.messages.FlinkJobNotFoundException: Could not find Flink job (d167f064a3e07d85f5050e5d6a5242db)]
Traceback (most recent call last):
  File "apache_beam/runners/common.py", line 961, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 554, in apache_beam.runners.common.SimpleInvoker.invoke_process
  File "/usr/local/lib/python3.7/site-packages/tfx/orchestration/beam/beam_dag_runner.py", line 84, in process
    self._run_component()
  File "/usr/local/lib/python3.7/site-packages/tfx/orchestration/beam/beam_dag_runner.py", line 88, in _run_component
    self._component_launcher.launch()
  File "/usr/local/lib/python3.7/site-packages/tfx/orchestration/launcher/base_component_launcher.py", line 205, in launch
    execution_decision.exec_properties)
  File "/usr/local/lib/python3.7/site-packages/tfx/orchestration/launcher/in_process_component_launcher.py", line 67, in _run_executor
    executor.Do(input_dict, output_dict, exec_properties)
  File "/usr/local/lib/python3.7/site-packages/tfx/components/example_gen/base_example_gen_executor.py", line 249, in Do
    artifact_utils.get_split_uri(output_dict['examples'], split_name)))
  File "/usr/local/lib/python3.7/site-packages/apache_beam/pipeline.py", line 524, in __exit__
    self.run().wait_until_finish()
  File "/usr/local/lib/python3.7/site-packages/apache_beam/pipeline.py", line 497, in run
    self._options).run(False)
  File "/usr/local/lib/python3.7/site-packages/apache_beam/pipeline.py", line 510, in run
    return self.runner.run_pipeline(self, self._options)
  File "/usr/local/lib/python3.7/site-packages/apache_beam/runners/portability/portable_runner.py", line 421, in run_pipeline
    result.wait_until_finish()
  File "/usr/local/lib/python3.7/site-packages/apache_beam/runners/portability/portable_runner.py", line 550, in wait_until_finish
    (self._job_id, self._state, self._last_error_message()))
RuntimeError: Pipeline BeamApp-root-0625125240-a1482bc6_0406fec0-114e-4cab-97d0-d4c3c9a72f9a failed in state FAILED: org.apache.flink.runtime.rest.util.RestClientException: [org.apache.flink.runtime.messages.FlinkJobNotFoundException: Could not find Flink job (d167f064a3e07d85f5050e5d6a5242db)]
```

And the jobserver log:
```
> Task :runners:flink:1.9:job-server:runShadowListening for transport dt_socket at address: 5005
[main] INFO org.apache.beam.runners.fnexecution.jobsubmission.JobServerDriver - LegacyArtifactStagingService started on localhost:8098[main] INFO org.apache.beam.runners.fnexecution.jobsubmission.JobServerDriver - Java ExpansionService started on localhost:8097[main] INFO org.apache.beam.runners.fnexecution.jobsubmission.JobServerDriver - JobService started on localhost:8099[grpc-default-executor-0] INFO org.apache.beam.runners.flink.FlinkJobInvoker - Invoking job BeamApp-root-0625125240-a1482bc6_0406fec0-114e-4cab-97d0-d4c3c9a72f9a with pipeline runner org.apache.beam.runners.flink.FlinkPipelineRunner@130b8c9d
[grpc-default-executor-0] INFO org.apache.beam.runners.fnexecution.jobsubmission.JobInvocation - Starting job invocation BeamApp-root-0625125240-a1482bc6_0406fec0-114e-4cab-97d0-d4c3c9a72f9a
[flink-runner-job-invoker] INFO org.apache.beam.runners.flink.FlinkPipelineRunner - Translating pipeline to Flink program.
[flink-runner-job-invoker] INFO org.apache.beam.runners.flink.FlinkExecutionEnvironments - Creating a Batch Execution Environment.
[flink-runner-job-invoker] INFO org.apache.beam.runners.flink.FlinkExecutionEnvironments - Using Flink Master URL localhost:8081.
[flink-runner-job-invoker] INFO org.apache.flink.api.java.ExecutionEnvironment - The job has 0 registered types and 0 default Kryo serializers[flink-runner-job-invoker] INFO org.apache.flink.configuration.Configuration - Config uses fallback configuration key 'jobmanager.rpc.address' instead of key 'rest.address'
[flink-runner-job-invoker] INFO org.apache.flink.runtime.rest.RestClient - Rest client endpoint started.
[flink-runner-job-invoker] INFO org.apache.flink.client.program.rest.RestClusterClient - Submitting job d167f064a3e07d85f5050e5d6a5242db (detached: false).
[flink-runner-job-invoker] INFO org.apache.flink.runtime.rest.RestClient - Shutting down rest endpoint.[flink-runner-job-invoker] INFO org.apache.flink.runtime.rest.RestClient - Rest endpoint shutdown complete.
[flink-runner-job-invoker] ERROR org.apache.beam.runners.fnexecution.jobsubmission.JobInvocation - Error during job invocation BeamApp-root-0625125240-a1482bc6_0406fec0-114e-4cab-97d0-d4c3c9a72f9a.
org.apache.flink.client.program.ProgramInvocationException: Could not retrieve the execution result. (JobID: d167f064a3e07d85f5050e5d6a5242db)
        at org.apache.flink.client.program.rest.RestClusterClient.submitJob(RestClusterClient.java:255)
        at org.apache.flink.client.program.ClusterClient.run(ClusterClient.java:338)
        at org.apache.flink.client.program.ClusterClient.run(ClusterClient.java:326)
        at org.apache.flink.client.program.ClusterClient.run(ClusterClient.java:301)        at org.apache.flink.client.RemoteExecutor.executePlanWithJars(RemoteExecutor.java:209)
        at org.apache.flink.client.RemoteExecutor.executePlan(RemoteExecutor.java:186)        at org.apache.flink.api.java.RemoteEnvironment.execute(RemoteEnvironment.java:173)
        at org.apache.beam.runners.flink.FlinkBatchPortablePipelineTranslator$BatchTranslationContext.execute(FlinkBatchPortablePipelineTranslator.java:194)        at org.apache.beam.runners.flink.FlinkPipelineRunner.runPipelineWithTranslator(FlinkPipelineRunner.java:116)        at org.apache.beam.runners.flink.FlinkPipelineRunner.run(FlinkPipelineRunner.java:83)
        at org.apache.beam.runners.fnexecution.jobsubmission.JobInvocation.runPipeline(JobInvocation.java:83)
        at org.apache.beam.vendor.guava.v26_0_jre.com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
        at org.apache.beam.vendor.guava.v26_0_jre.com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
        at org.apache.beam.vendor.guava.v26_0_jre.com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.flink.runtime.rest.util.RestClientException: [org.apache.flink.runtime.messages.FlinkJobNotFoundException: Could not find Flink job (d167f064a3e07d85f5050e5d6a5242db)]
        at org.apache.flink.runtime.rest.RestClient.parseResponse(RestClient.java:389)
        at org.apache.flink.runtime.rest.RestClient.lambda$submitRequest$3(RestClient.java:373)
        at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:966)        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:940)        at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:456)
        ... 3 more
[flink-runner-job-invoker] INFO org.apache.beam.runners.fnexecution.artifact.AbstractLegacyArtifactRetrievalService - Manifest at /tmp/beam-artifact-staging/job_eddd16ea-2d80-4f2e-a816-4622f197fb7a/MANIFEST has 2 artifact locations
```

**Comment by ibzib** [comment]
Can you check the Flink JobManager logs at `localhost:8081`?

**Comment by Vaskozl** [comment]
I've re-ran `setup_beam_on_flink.sh` with 2.21 and the cherry-picks in a clean buster image with the recommended pip dependencies and I seem to be getting slightly further. I've update the logs in my last post to reflect that (previously ran the script dirtily after trying 2.22.0 beam)

The logs on the flink interface on localhost:8081 (and schematics) appear to show that the CsvExampleGen component ran to completion:
```
2020-06-25 18:19:59,258 INFO  org.apache.flink.runtime.executiongraph.ExecutionGraph        - DataSink (DiscardingOutput) (4/4) (5ece18779d1ac70c06705bc443a1b614) switched from RUNNING to FINISHED.
2020-06-25 18:19:59,280 INFO  org.apache.flink.runtime.executiongraph.ExecutionGraph        - DataSink (DiscardingOutput) (3/4) (b613cbb4d381af8d48070c5bdc2aa791) switched from RUNNING to FINISHED.
2020-06-25 18:19:59,281 INFO  org.apache.flink.runtime.executiongraph.ExecutionGraph        - DataSink (…ndler: Checkpointing is not enabled for this job (61808f53592145d790d36977295380d2).
2020-06-25 18:30:07,842 ERROR org.apache.flink.runtime.rest.handler.job.checkpoints.CheckpointConfigHandler  - Exception occurred in REST handler: Checkpointing is not enabled for this job (61808f53592145d790d36977295380d2).
2020-06-25 18:30:07,843 ERROR org.apache.flink.runtime.rest.handler.job.checkpoints.CheckpointingStatisticsHandler  - Exception occurred in REST handler: Checkpointing has not been enabled.
```

Nevertheless both the pipeline and jobserver logs error with:
```
RuntimeError: Pipeline BeamApp-root-0625181843-a9e6c02d_fb28e54a-187a-4767-bbf8-4abd660d2541 failed in state FAILED: java.lang.ClassCastException: org.apache.beam.runners.core.metrics.MetricsContainerStepMap cannot be cast to org.apache.beam.runners.core.metrics.MetricsContainerStepMap
```

**Comment by Vaskozl** [comment]
I've managed to run the example with by using these two official images:
https://hub.docker.com/r/apache/beam_flink1.10_job_server/tags
https://hub.docker.com/_/flink?tab=tags

With flink 1.10 (scal 2.11) and beam 2.21.0. Running a jobserver, jobmanager and taskmanager container - all using host networking.

I've so far been unable to run the example with anything other than host networking and `LOOPBACK` as the environment.

I've tried using both the `DOCKER` environment by exposing the docker socket to the taskmanager, and the `EXTERNAL` by using a python-pool container, but both resulted in blank gRPC errors.

**Comment by ferryvg** [comment]
I have some issue

Pipeline options:
```
--runner=FlinkRunner,
--flink_master=*.*.*.*:8081
--environment_type=EXTERNAL
--environment_config=*.*.*.*:50000
--flink_version=1.10
```

Job server and worker pool `docker-compose.yml`:
```
version: '3.7'
services:
  job-server:
    image: apache/beam_flink1.10_job_server:2.27.0
    container_name: beam-job-server
    command: [
            "--artifacts-dir", "/opt/artifacts",
        ]
    volumes:
      - ./data/artifacts:/opt/artifacts
    network_mode: host

  python_sdk:
    image: apache/beam_python3.8_sdk:2.27.0
    container_name: beam-python-sdk
    command: --worker_pool
    network_mode: host
```

Flink taskexecutor log:
```
Java.io.FileNotFoundException: /tmp/beam-temp3g1emmsd/artifactsbv63p2is/a9d47b4ae5a1f4a0b5fddd51d2cd152583637e1353a615d5457aa19494051826/1-ref_Environment_default_e-tfx_ephemeral-0.27.0.tar.gz (No such file or directory)
        at java.io.FileInputStream.open0(Native Method)
        at java.io.FileInputStream.open(FileInputStream.java:195)
        at java.io.FileInputStream.<init>(FileInputStream.java:138)
        at org.apache.beam.sdk.io.LocalFileSystem.open(LocalFileSystem.java:127)
        at org.apache.beam.sdk.io.LocalFileSystem.open(LocalFileSystem.java:83)
        at org.apache.beam.sdk.io.FileSystems.open(FileSystems.java:256)
        at org.apache.beam.runners.fnexecution.artifact.ArtifactRetrievalService.getArtifact(ArtifactRetrievalService.java:124)
        at org.apache.beam.runners.fnexecution.artifact.ArtifactRetrievalService.getArtifact(ArtifactRetrievalService.java:99)
        at org.apache.beam.model.jobmanagement.v1.ArtifactRetrievalServiceGrpc$MethodHandlers.invoke(ArtifactRetrievalServiceGrpc.java:327)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:172)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.Contexts$ContextualizedServerCallListener.onHalfClose(Contexts.java:86)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:331)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:817)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at org.apache.beam.vendor.grpc.v1p26p0.io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

Beam worker pool log:
```
2021/02/10 12:28:10 Failed to retrieve staged files: failed to retrieve /tmp/staged in 3 attempts: failed to retrieve chunk for /tmp/staged/tfx_ephemeral-0.27.0.tar.gz
```

**Comment by ibzib** [comment]
@ferryvg your error is not exactly the same as the original post. Anyway, can you try adding `--flink_submit_uber_jar` to your pipeline options?

Also, when you use FlinkRunner (instead of PortableRunner), starting the job server as a separate service is unnecessary. I recommend removing the job server.

**Comment by ferryvg** [comment]
@ibzib Wow, this option is not documented... description is present only in source codes. Thanks

Yeah, i know about beam job server is not necessary, because i have Flink

**Comment by zhitaoli** [comment]
Sorry for the slow response.

Can you check whether https://github.com/tensorflow/tfx/commit/f50cea8d437de3d5908688b44c61c52f0f87a90f (available after tfx>=0.29) would fix the original issue?

**Comment by vaskozl** [comment]
Cheers @zhitaoli 

My original issue wasn't the exclusion of `uber_jar` as I was later able to get it running on beam 2.21.0 without the option. I had a hard time going from the example to a multi node setup. As I reported in an ealier, I was able to get it running by using the separate flink containers and host networking.

I then experienced it not working going from beam 2.21.0 to 2.27.0 which I think was fixed by the uber_jar flag as reported by ferryvg, but I haven't had time to test the flag myself and switch our pipelines back to flink.

Feel free to close the issue as it seems to have been solved.
